assets/src/dashboard/app/api/useSettingsApi.js

```diff
@@ -0,0 +1,118 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * External dependencies
+ */
+import { useCallback, useMemo, useReducer } from 'react';
+import queryString from 'query-string';
+
+/**
+ * Internal dependencies
+ */
+import settingsReducer, {
+  defaultSettingsState,
+  ACTION_TYPES as SETTINGS_ACTION_TYPES,
+} from '../reducer/settings';
+
+export default function useSettingsApi(
+  dataAdapter,
+  { globalStoriesSettingsApi }
+) {
+  const [state, dispatch] = useReducer(settingsReducer, defaultSettingsState);
+
+  const fetchGoogleAnalyticsId = useCallback(async () => {
+    if (!globalStoriesSettingsApi) {
+      dispatch({
+        type: SETTINGS_ACTION_TYPES.FETCH_GOOGLE_ANALYTICS_FAILURE,
+        payload: {
+          message: {
+            body: __('Cannot connect to data source', 'web-stories'),
+            title: __('Unable to find google analytics ID', 'web-stories'),
+          },
+        },
+      });
+    }
+    try {
+      const response = await dataAdapter.get(
+        queryString.stringifyUrl({
+          url: globalStoriesSettingsApi,
+        })
+      );
+
+      dispatch({
+        type: SETTINGS_ACTION_TYPES.FETCH_GOOGLE_ANALYTICS_SUCCESS,
+        payload: response.googleAnalyticsId,
+      });
+    } catch (err) {
+      dispatch({
+        type: SETTINGS_ACTION_TYPES.FETCH_GOOGLE_ANALYTICS_FAILURE,
+        payload: {
+          message: {
+            body: err.message,
+            title: __('Unable to find google analytics ID', 'web-stories'),
+          },
+        },
+      });
+    }
+  }, [dataAdapter, globalStoriesSettingsApi]);
+
+  const updateGoogleAnalyticsId = useCallback(
+    async (newAnalyticsId) => {
+      const query = { analyticsId: newAnalyticsId };
+
+      try {
+        const response = await dataAdapter.post(
+          queryString.stringifyUrl({
+            url: globalStoriesSettingsApi,
+            query,
+          })
+        );
+
+        dispatch({
+          type: SETTINGS_ACTION_TYPES.UPDATE_GOOGLE_ANALYTICS_SUCCESS,
+          payload: response.googleAnalyticsId,
+        });
+      } catch (err) {
+        dispatch({
+          type: SETTINGS_ACTION_TYPES.UPDATE_GOOGLE_ANALYTICS_FAILURE,
+          payload: {
+            message: {
+              body: err.message,
+              title: __('Unable to update google analytics ID', 'web-stories'),
+            },
+          },
+        });
+      }
+    },
+    [dataAdapter, globalStoriesSettingsApi]
+  );
+
+  const api = useMemo(
+    () => ({
+      fetchGoogleAnalyticsId,
+      updateGoogleAnalyticsId,
+    }),
+    [fetchGoogleAnalyticsId, updateGoogleAnalyticsId]
```

Comment:
Does the Google Analytics ID endpoint need to be its own separate thing?

If we were to add a second setting for the user to select a `Theme` (like dark or light or something like that), would we need `fetchTheme` and `updateTheme` functions too?  

Could we lump all the settings into one object and  just have `fetchUserSettings` and `updateUserSettings` that would return all settings and update any setting we passed in (including Google Analytics Id)?

Reply:
Pascal and I have been going back and forth about what this needs to be and he did a good job of fleshing out some implementation brief for this here: https://github.com/google/web-stories-wp/issues/3031 - take a look? 

Ideally this isn't just an API for google analytics, in my mind - that's why I named it `useSettingsApi` not `useGoogleAnalyticsApi`. I'm not sure what we'll need to pass back though, so if it'll be the same fetch and update or separate? I opted for saying fetchGoogleAnalyticsId and updateGoogleAnalyticsId for the time being to keep structure tight knowing that it's all placeholder. 

I don't mind updating to future proof it. Thoughts?

Reply:
Looking at the "Implementation Brief" in #3031, it says it'll register a new option `web_stories_ga_tracking_id` that will be exposed by the `Settings_Controller`.  And the `Settings_Controller` would be reached by the endpoint `web-stories/v1/settings`.

I interpret that to mean that there will be just one settings endpoint that will return the ga_tracking_id as part of its return settings object. 

I think your initial impulse to name it `useSettingsApi` is correct.  I'm just bumping against making all the other functions/actions/tests 'GoogleAnalytics' specific.  Looking at the brief it seems there will be one generic settings endpoint (`web-stories/v1/settings`) and it'll return an object that includes the `web_stories_ga_tracking_id`, so I think you should be able to keep the `useSettingsApi` completely generic and just have `fetchSettings` and `updateSettings` and treat the `ga_tracking_id` as merely a key/value pairing in the settings object.

Does that make sense?  

But you definitely have more insight into what's coming from the settings page, so if it doesn't make sense I'm happy to approve as is and should anything need to be updated once the API is actually working then it can be handled then.

Reply:
I can get behind that :) Given that update, i'm going to probably update the initial call for settings data to be in the parent of google analytics view.